### PR TITLE
[Launcher] mod manager drag'n'drop

### LINKED ISSUE #853: install mods by drag'n'drop of mod directory onto the launcher
in the future we can extend this to also drag'n'drop an archive

### PR DESCRIPTION
**Pull request by Laserlicht**
fixes #853

allows to drag zip files (e.g. manual downloads from github) to mod list and install the zip file.

also allows to install maps and campaigns with drag'n'drop.

### CODE REVIEW

File: launcher/modManager/cmodlistview_moc.cpp [hunk 1]
```diff
@@ -48,6 +48,46 @@ void CModListView::changeEvent(QEvent *event)
...
+ 
+void CModListView::dragMoveEvent(QDragMoveEvent* event)
+{
+	event->acceptProposedAction();
+}
+ 
+void CModListView::dragLeaveEvent(QDragLeaveEvent* event)
+{
+	event->accept();
+}
```
Comment (kambala-decapitator) [inline launcher/modManager/cmodlistview_moc.cpp:64]: Iirc these are not required to be implemented

File: launcher/modManager/cmodlistview_moc.cpp [hunk 2]
```diff
@@ -48,6 +48,46 @@ void CModListView::changeEvent(QEvent *event)
...
+void CModListView::dropEvent(QDropEvent* event)
+{
+	const QMimeData* mimeData = event->mimeData();
+
+	if(mimeData->hasUrls())
+	{
+		QStringList pathList;
+		QList<QUrl> urlList = mimeData->urls();
+
+		for (int i = 0; i < urlList.size(); i++)
```
Comment (kambala-decapitator) [inline launcher/modManager/cmodlistview_moc.cpp:75]: make urlList const and use for-in

File: launcher/modManager/cmodlistview_moc.cpp [hunk 3]
```diff
@@ -48,6 +48,46 @@ void CModListView::changeEvent(QEvent *event)
...
+	event->accept();
+}
+
+void CModListView::dropEvent(QDropEvent* event)
+{
+	const QMimeData* mimeData = event->mimeData();
+
+	if(mimeData->hasUrls())
+	{
+		QStringList pathList;
```
Comment (kambala-decapitator) [inline launcher/modManager/cmodlistview_moc.cpp:72]: Looks unused

File: launcher/modManager/cmodlistview_moc.cpp [hunk 4]
```diff
@@ -48,6 +48,46 @@ void CModListView::changeEvent(QEvent *event)
...
+
+		for (int i = 0; i < urlList.size(); i++)
+		{
+			QString url = urlList.at(i).toString();
+			if(url.endsWith(".zip", Qt::CaseInsensitive))
+				downloadFile(url.split("/").last().toLower()
+					// mod name currently comes from zip file -> remove suffixes from github zip download
+					.replace(QRegularExpression("-[0-9a-f]{40}"), "")
+					.replace(QRegularExpression("-vcmi-.*\\.zip"), ".zip")
+					.replace(QRegularExpression("-main.zip"), ".zip")
```
Comment (kambala-decapitator) [inline launcher/modManager/cmodlistview_moc.cpp:83]: Regex not needed on this line

File: launcher/modManager/cmodlistview_moc.cpp [hunk 5]
```diff
@@ -48,6 +48,46 @@ void CModListView::changeEvent(QEvent *event)
...
+		QList<QUrl> urlList = mimeData->urls();
+
+		for (int i = 0; i < urlList.size(); i++)
+		{
+			QString url = urlList.at(i).toString();
+			if(url.endsWith(".zip", Qt::CaseInsensitive))
+				downloadFile(url.split("/").last().toLower()
+					// mod name currently comes from zip file -> remove suffixes from github zip download
+					.replace(QRegularExpression("-[0-9a-f]{40}"), "")
+					.replace(QRegularExpression("-vcmi-.*\\.zip"), ".zip")
```
Comment (kambala-decapitator) [inline launcher/modManager/cmodlistview_moc.cpp:82]: .+ probably?

File: launcher/modManager/cmodlistview_moc.cpp [hunk 6]
```diff
@@ -48,6 +48,46 @@ void CModListView::changeEvent(QEvent *event)
...
+					.replace(QRegularExpression("-main.zip"), ".zip")
+					, url, "mods", 0);
+			else
+				downloadFile(url.split("/").last(), url, "mods", 0);
```
Comment (kambala-decapitator) [inline launcher/modManager/cmodlistview_moc.cpp:86]: Getting last component is copy-pasted

Also, doesn't QUrl have method to get last component?

File: launcher/modManager/cmodlistview_moc.cpp [hunk 7]
```diff
@@ -685,6 +728,8 @@ void CModListView::installFiles(QStringList files)
 	{
 		if(filename.endsWith(".zip"))
 			mods.push_back(filename);
+		if(filename.endsWith(".h3m") || filename.endsWith(".h3c") || filename.endsWith(".vmap") || filename.endsWith(".vcmp"))
```
Comment (kambala-decapitator) [inline launcher/modManager/cmodlistview_moc.cpp:731]: Use else if here and on the next condition

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

File: launcher/modManager/cmodlistview_moc.cpp (hunk 1 above)
Comment (Laserlicht) [reply launcher/modManager/cmodlistview_moc.cpp]: dragEnterEvent is needed

File: launcher/modManager/cmodlistview_moc.cpp [hunk 8]
```diff
@@ -685,7 +718,9 @@ void CModListView::installFiles(QStringList files)
 	{
 		if(filename.endsWith(".zip"))
 			mods.push_back(filename);
-		if(filename.endsWith(".json"))
+		else if(filename.endsWith(".h3m") || filename.endsWith(".h3c") || filename.endsWith(".vmap") || filename.endsWith(".vcmp"))
+			maps.push_back(filename);
+		else if(filename.endsWith(".json"))
```
Comment (IvanSavenko) [inline launcher/modManager/cmodlistview_moc.cpp:723]: According to Qt docs these checks are case-sensitive. Worked fine for "nice" data from github, but user-provided files might come with any case.

File: launcher/modManager/cmodlistview_moc.cpp [hunk 9]
```diff
@@ -794,6 +832,16 @@ void CModListView::installMods(QStringList archives)
 		QFile::remove(archive);
 }
 
+void CModListView::installMaps(QStringList archives)
```
Comment (IvanSavenko) [inline launcher/modManager/cmodlistview_moc.cpp:835]: These are maps, not "archives"?

File: launcher/modManager/cmodlistview_moc.cpp [hunk 10]
```diff
@@ -48,6 +48,36 @@ void CModListView::changeEvent(QEvent *event)
...
+
+void CModListView::dropEvent(QDropEvent* event)
+{
+	const QMimeData* mimeData = event->mimeData();
+
+	if(mimeData->hasUrls())
+	{
+		const QList<QUrl> urlList = mimeData->urls();
+
+		for (auto & url : urlList)
```
Comment (kambala-decapitator) [inline launcher/modManager/cmodlistview_moc.cpp:64]:
```suggestion
		for (const auto & url : urlList)
```

**Review by kambala-decapitator — APPROVED** [review]

File: launcher/modManager/cmodlistview_moc.cpp [hunk 11]
```diff
@@ -48,6 +48,36 @@ void CModListView::changeEvent(QEvent *event)
 	QWidget::changeEvent(event);
 }
 
+void CModListView::dragEnterEvent(QDragEnterEvent* event)
+{
+	event->acceptProposedAction();
+}
```
Comment (IvanSavenko) [inline launcher/modManager/cmodlistview_moc.cpp:54]: According to Qt docs we should indicate whether we accept event:
```
 void Window::dragEnterEvent(QDragEnterEvent *event)
 {
     if (event->mimeData()->hasFormat("text/plain"))
         event->acceptProposedAction();
 }
```
( I guess in our case we should check `mimeData->hasUrls()` and look up whether there are any maps/zips/etc)

**Pull request merged**